### PR TITLE
docs: Add initial docs on how to use node local storage

### PR DESCRIPTION
**Pull request by rata**
This adds basic documentation on how to use node local storage using
local persistent volumes in filesystem mode.

Other sections are mentioned, although empty for now, on other ways to
use it that we may want to expand later.

### CODE REVIEW

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: I would capitalize it (except "on")

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: `+` storage

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: capital L (also below)

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: I would call it "Local Persistent Volumes" because it's used as name for a concept

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: I would write "Persistent Volume". (The Kubernetes docs even write it as "`PersistentVolume` (PV)".)

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: I think no "(link)" is needed if it's the section below – or which one is meant?

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: captial K

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: It would be good to make it clickable when rendered as markdown.

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: `+` which …

File: docs/advanced/node-local-storage.md
Comment (pothos) [inline docs/advanced/node-local-storage.md]: "StatefulSet", "stateful set", or "Stateful Set"? Just choose one style, I would tend to "Stateful Set" if we talk about it as concept and not as configuration part.

File: docs/advanced/node-local-storage.md
Comment (invidian) [reply docs/advanced/node-local-storage.md]: Use https://capitalizemytitle.com/ :)

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
bare-metal storage performance inside Kubernetes pods.
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
This guide shows different ways to use node local storage on Lokomotive, to achieve near
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
# Using Node Local Storage on Kubernetes
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
to use one persistent volume in the host. However, see the section(link) below
```

I think "link" might be a placeholder for anchor to section below. Still needs to be added. Or just remove it and have `"section below".

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]: This sentence doesn't make sense to me.

>More details about specific volumes used

used?

```suggestion
More details about specific volumes used and how to use them on Lokomotive will
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
follow, so there is no need to look into volume specifics on the first read.
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
**NOTE**: This has been tested on Packet provider only, some Lokomotive changes may need to be
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
ported to other platforms for this to work.
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]: How about:
```suggestion
## Local Persistent Volumes in Filesystem Mode
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
The Kubernetes documentation on Local Persistent Volumes using filesystem mode is
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
There are some limitations you might want to have in mind:
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
devices on `/mnt/sda` or paths inside `/mnt` at your convenience.
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
This is the only automated setup supported in lokomotive at the time of writing.
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
   to use different one right now.
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]: I think comment about PR is not really needed here.

>uber experience

uber? What does it mean?
```suggestion
else and some useful tips from uber experience
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
## Local Persistent Volumes in Raw Block Device Mode
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]: It's called `OpenEBS Local PV` upstream: https://docs.openebs.io/docs/next/uglocalpv.html
```suggestion
## Using OpenEBS Local PV
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
We might consider other container attached storage (CAS) solutions, like [Rook](https://rook.io/),
```

**Review by invidian — CHANGES_REQUESTED** [review]
Added suggestions. The overall document seems all right.

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: True, no link. It was a leftover. Thanks :)

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Uber the company, sorry I wasn't clear :D

Why do you think is not needed? Because we can reference the blog post or because it does not adds no value knowing that in this document?

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: I wouldn't, but if all agree (and looking quickly it seems that is done in another docs), so did it :)

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: What I wanted to say is that read only the generalities and not everything about every volume possible (like Local, NFS, etc.) and more details about the types of volume used in the guide will follow.

Would something like:

```
More detail about specific types of volumes used and how to use them on Lokomotive will
follow, so no need to look into volume specifics on the first read.
```
make more sense to you?

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Why change `at the time of this writing` by `at the time of writing`?

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Fixed, thanks!

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Thanks! Did a s/lokomotive/Lokomotive (excluding URLs :))

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Thanks, fixed!

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Makes sense, changed to capital in all the docoumen. Thanks! :)

File: docs/advanced/node-local-storage.md
Comment (invidian) [reply docs/advanced/node-local-storage.md]: >Uber the company, sorry I wasn't clear :D

Maybe `Uber's experience` then?

>Why do you think is not needed? Because we can reference the blog post or because it does not adds no value knowing that in this document?

I would create an issue instead if you would like to file a PR to Kubernetes docs, so we don't forget about it.

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Thanks, changed it in the whole doc :)

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Thanks!

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Oh, yes! That was an internal note and I will either create an issue or a PR for that. Is that whan you think was not needed? I obviosly agree, that was a mistake :)

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Good point, changed it all to "Stateful Set". Thanks!

File: docs/advanced/node-local-storage.md
Comment (invidian) [reply docs/advanced/node-local-storage.md]: https://english.stackexchange.com/questions/199381/is-at-the-time-of-writing-correct `at the time of writing` is the better version of this phrase as far as I know.

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Removed, it was a placeholder but removing is better. Thanks!

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Right! Improved all the links now, thanks!

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Oh, thanks!

**Comment by rata** [comment]
@pothos @invidian thanks a lot for the detailed review! And the commit suggestions!

I think it is much improved now (added all your suggestions :)), PTAL @pothos @invidian :)

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]: Too many `one`s here it seems. How about:
```suggestion
Single Persistent Volume can be bound to only one PVC, so only one pod will be able
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]: Too many `one`s here it seems. How about:
```suggestion
to use this Persistent Volume in the host. However, see the section below to see
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
In this example 2 nodes with Local Persistent Volumes will be created and used
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
### Automatic Local Persistent Volume Creation and Handling
```

File: docs/advanced/node-local-storage.md
Comment (invidian) [inline docs/advanced/node-local-storage.md]:
```suggestion
### Useful Links
```

**Review by invidian — DISMISSED** [review]
Few more found.

File: docs/advanced/node-local-storage.md
Comment (surajssd) [reply docs/advanced/node-local-storage.md]: Please link to that sub section?

File: docs/advanced/node-local-storage.md
Comment (surajssd) [inline docs/advanced/node-local-storage.md:204]: nit: in a single code block like this if there is a empty line separating multiple commands and it's output it will be helpful.

File: docs/advanced/node-local-storage.md
Comment (surajssd) [inline docs/advanced/node-local-storage.md]: This section deserves to be bold, or heading or something or could go into a FAQ section of its own.

File: docs/advanced/node-local-storage.md
Comment (surajssd) [inline docs/advanced/node-local-storage.md]: I am not sure if rest of this document should be here, rather than helping user it is just confusing user to follow lot of other links.

I would rather let this document be only something we support and have tried, tested and trusted.

File: docs/advanced/node-local-storage.md
Comment (surajssd) [inline docs/advanced/node-local-storage.md:341]: Like I said earlier in the previous comment this deserves its own document.

**Review by surajssd — COMMENTED** [review]
Few nits and suggestions. The document is helpful. But the last section can go into other doc or be totally removed IMO. With so many links and lot of text it is just confusing than of help to users.

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: s/Single/A single/?

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md:204]: Thanks, fixed!

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md:341]: I prefer to start with a single document of node local storage, if at a later stage we want to have more docs (1 doc per section, a brief summary on each section and then link to another doc, etc.) we can do when we add more stuff.

 What do you think?

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Not sure to create one section just for one "what if" paragraph... Maybe a sub section in the example?

Comment (surajssd) [reply docs/advanced/node-local-storage.md]: sure

File: docs/advanced/node-local-storage.md
Comment (surajssd) [reply docs/advanced/node-local-storage.md:341]: I want to have a simple and consice documentation for the users so that we can just point them to do certain task. Like for manual node local PV I would point someone to this doc. The doc should not confuse user.

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Okay... But I think that is useful and I would prefer to have this somewhere, where do you suggest? Or you suggest to just drop this?

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md:341]: Can you please elaborate on why would this confuse the user? I only see an almost empty section that says "we will expand on this later, these links might be relevant for that".

Would an empty section or removing this be better from your point of view? I'm not sure what to do to address your comment :-/. An new doc with this links doesn't seem great (it's pretty much empty) and not sure what other ways are to address this.

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]: Marked bold, thanks!

**Comment by rata** [comment]
@surajssd PTAL. I've addressed the all comments, except the ones I asked you something, and adapted to use `/mnt/node-local-storage` as path, as that was recently changed in master

File: docs/advanced/node-local-storage.md
Comment (kosyfrances) [inline docs/advanced/node-local-storage.md]: s/volumes types/volume types

File: docs/advanced/node-local-storage.md
Comment (kosyfrances) [inline docs/advanced/node-local-storage.md]: s/in/on

File: docs/advanced/node-local-storage.md
Comment (kosyfrances) [inline docs/advanced/node-local-storage.md:27]: How about rephrasing as:

```
See the section below for information on how to create more than one Persistent
Volume per node.
```

File: docs/advanced/node-local-storage.md
Comment (kosyfrances) [inline docs/advanced/node-local-storage.md]: Do you mean that the block mode will be covered in a different section of this document? If yes, how about rephrasing as:

```
Note that this section is about the filesystem mode, not the block mode.
Information about the block mode will be covered on a different section of this document.
```

File: docs/advanced/node-local-storage.md
Comment (kosyfrances) [inline docs/advanced/node-local-storage.md]: Add a comma after `example`.

Change `2` to `two` to be consistent with the remaining part of the sentence.

File: docs/advanced/node-local-storage.md
Comment (kosyfrances) [inline docs/advanced/node-local-storage.md]: Do you mean to say: 
```
Note that when using the `Delete` .....
```

File: docs/advanced/node-local-storage.md
Comment (kosyfrances) [inline docs/advanced/node-local-storage.md]:
```
so you may want to choose the policy that better fits
```
--->
```
so you may want to choose a policy that better fits
```

**Review by kosyfrances — COMMENTED** [review]
This is what I have reviewed for now, will continue another time.

File: docs/advanced/node-local-storage.md
Comment (kosyfrances) [inline docs/advanced/node-local-storage.md]: No you won't see it on the host, although if you delete the pod and a new one gets created, it will still have the same contents of the old pod. 

I did a setup using this PR, but whatever I wrote to the pod on `/usr/test-pod`, did not show up on the host's `/mnt/node-local-storage` :confused: . So, I decided to try with a different directory. I created a `/mnt/data` directory on the host and it had the following permissions by default.
```
core@kosy-worker-worker-0 /mnt $ ls -l
total 12
drwxr-xr-x. 2 root root 4096 Sep 23 09:59 data
drwxr-xr-x. 3 root root 4096 Sep 23 07:33 node-local-storage
```
After creating the pv to use `/mnt/data` instead of `node-local-storage`, I wrote to the pod and saw the new file also in `/mnt/data` and noticed that the permissions on `/mnt/data/` changed by itself to:
```
core@kosy-worker-worker-0 /mnt $ ls -ll
total 12
drwxrwsr-x. 2 root bin  4096 Sep 23 10:15 data
drwxr-xr-x. 3 root root 4096 Sep 23 07:33 node-local-storage
```

**Comment by rata** [comment]
Node local storage was moved back to `/mnt` in this PR https://github.com/kinvolk/lokomotive-kubernetes/pull/72 recently merged, so we need to adjust this doc to reflect that

**Comment by invidian** [comment]
@rata do you have time to take care of this PR?

**Comment by rata** [comment]
@invidian not really, and is kind of outdated now. I'll try to keep an eye on this during next week, but no promises :(

**Comment by invidian** [comment]
I'm taking a look into it to get it merged.

File: docs/advanced/node-local-storage.md
Comment (rata) [inline docs/advanced/node-local-storage.md]: This was not the case before self-hosted kubelet, as the pod didn't saw mounts inside mnt (see https://github.com/kinvolk/lokomotive-kubernetes/issues/73) Not sure if it is now possible with self-hosted kubelet

**Comment by rata** [comment]
IMHO, it looks fine besides the two open comments (the one I just posted and suraj comment from long ago).

Maybe we can oing and wait for suraj to comment again?

File: docs/advanced/node-local-storage.md
Comment (rata) [reply docs/advanced/node-local-storage.md]:
```suggestion
Lokomotive Kubernetes currently only exposes `/mnt/` to the kubelet, so only that
directory can be used for node local storage.
```

Comment (rata) [inline docs/advanced/node-local-storage.md]:
```suggestion

1. Lokomotive Kubernetes currently only exposes `/mnt/` to the kubelet, so only that
directory can be used for node local storage.
```

Comment (rata) [reply docs/advanced/node-local-storage.md]: If' say remove the "You can **manually** mount different devices on `/mnt/sda` or paths inside `/mnt` at your convenience" but I'm failing to create a github suggestion for this, not sure why :man_facepalming:

File: docs/advanced/node-local-storage.md
Comment (rata) [inline docs/advanced/node-local-storage.md]: For the same issue (https://github.com/kinvolk/lokomotive-kubernetes/issues/73) that was discovered later, I think this _might_ not be accurate either. I'd remove this, probably. What do you think?

The link (the `[1]`), is also shared in the section below, so we still keep a reference for that useful doc :)

Comment (invidian) [reply docs/advanced/node-local-storage.md]: I guess it would work, if we enable mount propagation for this path, which we don't at the moment. I can remove it then.

File: docs/advanced/node-local-storage.md
Comment (invidian) [reply docs/advanced/node-local-storage.md:341]: @surajssd can you look into the modified version? I think it's fine for now to keep it here until we have something more specific. Then we can just refer here to new document.

**Review by surajssd — DISMISSED** [review]
This looks ok.

File: docs/advanced/node-local-storage.md
Comment (invidian) [reply docs/advanced/node-local-storage.md]: Ok, removed now.

**Comment by rata** [comment]
@surajssd do you think it looks good to approve the PR? Can you approve it, then? :)

@invidian can you approve too? (not sure, it's weird, you addressed the last comments at least. But I can't approve as it is my PR :))

@pothos or maybe can you take a look and approve if it seems okay? :)

PS: I'd merge with squash and merge, if we are all ok with this (don't see value in the split commits)

**Comment by invidian** [comment]
>@surajssd do you think it looks good to approve the PR? Can you approve it, then? :)

@surajssd already approved it, but it got dismissed by my push. Let me squash the commits then.

**Review by invidian — APPROVED** [review]
Approving on behalf of @rata

**Comment by rata** [comment]
Merging on behalf of @invidian ? :joy:

**Comment by rata** [comment]
@invidian thanks **a lot** for driving this forward! :)

**Pull request merged**
